File: backend/app/schemas/course.py
from typing import Optional, List, Any
from pydantic import BaseModel
from datetime import datetime
from app.models.enums import LevelEnum, CategoryEnum


class CourseBase(BaseModel):
    title: str
    description: Optional[str] = None
    thumbnail: Optional[str] = None
    level: LevelEnum = LevelEnum.beginner
    category: CategoryEnum = CategoryEnum.paid
    price: Optional[float] = 0.0
    tags: Optional[List[str]] = []


class CourseCreate(CourseBase):
    pass


class CourseUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    thumbnail: Optional[str] = None
    level: Optional[LevelEnum] = None
    category: Optional[CategoryEnum] = None
    price: Optional[float] = None
    is_published: Optional[bool] = None
    tags: Optional[List[str]] = None


class CourseResponse(CourseBase):
    id: int
    created_by: int
    is_published: bool
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class InstructorSchema(BaseModel):
    id: int
    full_name: str
    photo: Optional[str] = None
    bio: Optional[str] = None
    
    class Config:
        from_attributes = True


class CourseDetailResponse(CourseResponse):
    instructors: List[InstructorSchema] = []

File: frontend/src/types/course.ts
export const LevelEnum = {
    BEGINNER: "beginner",
    INTERMEDIATE: "intermediate",
    ADVANCED: "advanced",
} as const;
export type LevelEnum = typeof LevelEnum[keyof typeof LevelEnum];

export const CategoryEnum = {
    FREE: "free",
    PAID: "paid",
} as const;
export type CategoryEnum = typeof CategoryEnum[keyof typeof CategoryEnum];

export interface Course {
    id: number;
    title: string;
    description?: string;
    thumbnail?: string;
    level: LevelEnum;
    category: CategoryEnum;
    price?: number;
    tags?: string[];
    duration_weeks?: number;
    created_by: number;
    is_published: boolean;
    created_at?: string;
    updated_at?: string;
    enrollments_count?: number;
}

export interface CourseCreate {
    title: string;
    description?: string;
    thumbnail?: string;
    level: LevelEnum;
    category: CategoryEnum;
    price?: number;
    tags?: string[];
}

export interface CourseUpdate {
    title?: string;
    description?: string;
    thumbnail?: string;
    level?: LevelEnum;
    category?: CategoryEnum;
    price?: number;
    is_published?: boolean;
    tags?: string[];
}

export interface CourseFilters {
    search?: string;
    category?: CategoryEnum;
    level?: LevelEnum;
    skip?: number;
    limit?: number;
}
